content/blog/posts/blogpost-motion-matching.md

```diff
@@ -0,0 +1,189 @@
+
+Using the query vector, we can find the next best matching frame in the motion database and start transitioning towards that.
+
+In case the new best matching frame candidate is close to the time in the animation that we are already playing, we don't do anything as we seem to be at the sweet spot in the motion database already.
+
+Pseudo-code:
+```
+// Keep playing the current animation.
+currentMotion.Update(timeDelta);
+ 
+if (Is it time to search for a new best matching frame?) // We might e.g. do this 5x a second
+{
+    // Evaluate the current pose including joint velocities.
+    queryPose = SamplePose(newMotionTime);
+ 
+    // Update the input query vector (Calculate features for the query pose)
+    queryValues = CalculateFeaturesFromPose(queryPose);
+ 
+    // Find the frame with the lowest cost based on the query vector.
+    bestMatchingFrame = FindBestMatchingFrame(queryValues);
+ 
+    // Start transitioning towards the new best matching frame in case it is not
+    // really close to the frame we are already playing.
+    if (IsClose(bestMatchingFrame, currentMotion.GetFrame()) == false)
+    {
+        StartTransition(bestMatchingFrame);
+    }
+}
+```
+
+#### Cost function
+
+The core question the algorithm tries to answer is: Where do we jump and transition to? The algorithm tries to find the best time in the motion database that matches the current character pose including its movements and the user input. To compare the frame candidates with each other, we use a cost function.
+
+The feature schema defines what the cost function looks like. Every feature added to the feature schema adds up to the cost. The bigger the discrepancy between the current velocity and the one from the frame candidate, the higher the penalty to the cost and the less likely the candidate is a good one to take.
+
+This makes motion matching an optimization problem where the frame with the minimum cost is the most preferred candidate to transition to.
+
+#### Searching next best matching frame
+The actual search happens in two phases, a broad phase to eliminate most of the candidates followed by a narrow phase to find the actual best candidate.
+
+##### 1. Broad-phase (KD-tree)
+A KD-tree is used to find the nearest neighbors (frames in the motion database) to the query vector (given input). The result is a set of pre-selected frames for the next best matching frame that is passed on to the narrow-phase. By adjusting the maximum tree depth or the minimum number of frames for the leaf nodes, the resulting number of frames can be adjusted. The bigger the set of frames the broad-phase returns, the more candidates the narrow-phase can choose from, the better the visual quality of the animation but the slower the algorithm.
```

Comment:
nit: It's a good idea to be consistent here and have line breaks after each heading (anything using `#`)